File: executor/executor/github.py
from .utils import log
import jwt
import requests
import threading
import time


class GitHub:
    def __init__(self, cli):
        self.org = cli.github_org

        self._http = requests.Session()
        self._http.headers["User-Agent"] = (
            "rust-lang/gha-self-hosted (infra@rust-lang.org)"
        )

        log(f"generating a JWT to authenticate as app {cli.github_client_id}")
        bearer = jwt.encode(
            {
                "iat": int(time.time() - 60),
                "exp": int(time.time() + 60 * 5),
                "iss": cli.github_client_id,
            },
            open(cli.github_private_key, "rb").read(),
            algorithm="RS256",
        )

        log(f"retrieving app installation id for {self.org}")
        resp = self._handle_error(
            self._http.get(
                f"https://api.github.com/orgs/{self.org}/installation",
                headers={"Authorization": f"Bearer {bearer}"},
            )
        )
        installation = resp.json()["id"]

        log(f"retrieving token for installation {installation}")
        resp = self._handle_error(
            self._http.post(
                f"https://api.github.com/app/installations/{installation}/access_tokens",
                headers={"Authorization": f"Bearer {bearer}"},
            )
        )

        self._http.headers["Authorization"] = f"token {resp.json()['token']}"

    def _handle_error(self, response: requests.Response) -> requests.Response:
        if response.status_code >= 400:
            print(
                f"error: github responded with status {response.status_code} to the request"
            )
            print(f"url: {response.url}")
            print(f"message: {response.json()['message']}")
            exit(1)
        return response

    def fetch_registration_token(self) -> str:
        log(f"fetching the GHA installation token for {self.org}")
        resp = self._handle_error(
            self._http.post(
                f"https://api.github.com/orgs/{self.org}/actions/runners/registration-token"
            )
        )
        return resp.json()["token"]


# How many seconds should pass between each call to the GitHub API.
GITHUB_API_POLL_INTERVAL = 15


class GitHubRunnerStatusWatcher(threading.Thread):
    def __init__(self, gh, repo, runner_name, check_interval, then):
        super().__init__(name="github-runner-status-watcher", daemon=True)

        self._check_interval = check_interval
        self._gh = gh
        self._repo = repo
        self._runner_name = runner_name
        self._then = then

    def run(self):
        log("started polling GitHub to detect when the runner started working")
        while True:
            runners = self._retrieve_runners()
            if self._runner_name in runners and runners[self._runner_name]["busy"]:
                log("the runner started processing a build!")
                self._then()
                break
            time.sleep(self._check_interval)

    def _retrieve_runners(self):
        result = {}
        url = f"https://api.github.com/repos/{self._repo}/actions/runners"
        # TODO: this is broken
        for response in self._gh._handle_error(self._gh._http.get(url)).json():
            for runner in response["runners"]:
                result[runner["name"]] = runner
        return result

File: executor/src/github.rs
use anyhow::{Context, Error, bail};
use jsonwebtoken::{Algorithm, EncodingKey};
use reqwest::Method;
use reqwest::{Client, RequestBuilder, Response};
use std::path::{Path, PathBuf};
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};
use tokio::time::sleep;

const REFRESH_TOKEN_AFTER: Duration = Duration::from_secs(60 * 15);
const POLL_INTERVAL: Duration = Duration::from_secs(5);

#[derive(Clone)]
pub struct GitHub {
    inner: Arc<GitHubInner>,
}

impl GitHub {
    pub fn new(client_id: &str, private_key: &Path, org: &str) -> Result<Self, Error> {
        let http = Client::builder()
            .user_agent("rust-lang/gha-self-hosted (infra@rust-lang.org)")
            .build()?;

        Ok(Self {
            inner: Arc::new(GitHubInner {
                http,
                org: org.into(),
                client_id: client_id.into(),
                private_key: private_key.into(),
                token: Mutex::new(None),
            }),
        })
    }

    pub async fn run_after_build_started(&self, id: u64, then: impl FnOnce()) {
        let mut last_status = None;
        let mut not_yet_executed = Some(then);

        loop {
            match self.get_runner(id).await {
                Ok(runner) => {
                    if last_status.as_ref() != Some(&runner.status) {
                        last_status = Some(runner.status.clone());
                        eprintln!("runner status changed to {}", runner.status);
                    }
                    if runner.busy
                        && let Some(callback) = not_yet_executed.take()
                    {
                        eprintln!("the runner started processing a build");
                        callback();
                        break;
                    }
                }
                Err(err) => {
                    eprintln!("warning: failed to poll GitHub for the runner status: {err}");
                }
            };
            sleep(POLL_INTERVAL).await;
        }
    }

    pub async fn create_runner(
        &self,
        name: &str,
        runner_group_id: u64,
        labels: &[&str],
    ) -> Result<CreatedRunner, Error> {
        #[derive(serde::Serialize)]
        struct RunnerBody<'a> {
            name: &'a str,
            labels: &'a [&'a str],
            runner_group_id: u64,
        }
        #[derive(serde::Deserialize)]
        struct RunnerResponse {
            runner: RunnerRunner,
            encoded_jit_config: String,
        }
        #[derive(serde::Deserialize)]
        struct RunnerRunner {
            id: u64,
        }

        let response: RunnerResponse = self
            .request(
                Method::POST,
                AuthMode::Token,
                &format!(
                    "https://api.github.com/orgs/{}/actions/runners/generate-jitconfig",
                    self.inner.org
                ),
            )
            .await?
            .json(&RunnerBody {
                name,
                runner_group_id,
                labels,
            })
            .send()
            .await?
            .handle_github_error()
            .await?
            .json()
            .await?;

        Ok(CreatedRunner {
            id: response.runner.id,
            jitconfig: response.encoded_jit_config,
        })
    }

    pub async fn get_runner(&self, id: u64) -> Result<RunnerInfo, Error> {
        let response = self
            .request(
                Method::GET,
                AuthMode::Token,
                &format!(
                    "https://api.github.com/orgs/{}/actions/runners/{id}",
                    self.inner.org
                ),
            )
            .await?
            .send()
            .await?
            .handle_github_error()
            .await?
            .json()
            .await?;
        Ok(response)
    }

    async fn refresh_token(&self) -> Result<String, Error> {
        #[derive(serde::Deserialize)]
        struct Installation {
            id: u64,
        }
        #[derive(serde::Deserialize)]
        struct Token {
            token: String,
        }

        eprintln!("refreshing the token for organization {}", self.inner.org);

        let jwt = self.generate_jwt()?;
        let installation = self
            .request(
                Method::GET,
                AuthMode::Jwt(&jwt),
                &format!(
                    "https://api.github.com/orgs/{}/installation",
                    self.inner.org
                ),
            )
            .await?
            .send()
            .await?
            .handle_github_error()
            .await?
            .json::<Installation>()
            .await?
            .id;
        let token = self
            .request(
                Method::POST,
                AuthMode::Jwt(&jwt),
                &format!("https://api.github.com/app/installations/{installation}/access_tokens"),
            )
            .await?
            .send()
            .await?
            .handle_github_error()
            .await?
            .json::<Token>()
            .await?
            .token;

        *self.inner.token.lock().unwrap() = Some(GitHubToken {
            token: token.clone(),
            issued_at: Instant::now(),
        });

        Ok(token)
    }

    fn generate_jwt(&self) -> Result<String, Error> {
        #[derive(serde::Serialize)]
        struct Claims<'a> {
            iat: u64,
            exp: u64,
            iss: &'a str,
        }

        let now = SystemTime::now().duration_since(UNIX_EPOCH)?.as_secs();

        Ok(jsonwebtoken::encode(
            &jsonwebtoken::Header {
                alg: Algorithm::RS256,
                ..Default::default()
            },
            &Claims {
                iat: now - 60,
                exp: now + 60 * 5,
                iss: &self.inner.client_id,
            },
            &EncodingKey::from_rsa_pem(&std::fs::read(&self.inner.private_key)?)?,
        )?)
    }

    async fn request(
        &self,
        method: Method,
        auth: AuthMode<'_>,
        url: &str,
    ) -> Result<RequestBuilder, Error> {
        let mut request = self.inner.http.request(method, url);
        match auth {
            AuthMode::Jwt(jwt) => {
                request = request.header("authorization", format!("Bearer {jwt}"));
            }
            AuthMode::Token => {
                let cache = self.inner.token.lock().unwrap().clone();
                let token = match cache {
                    Some(t) if t.issued_at.elapsed() > REFRESH_TOKEN_AFTER => {
                        Box::pin(self.refresh_token())
                            .await
                            .context("failed to refresh token")?
                    }
                    Some(t) => t.token.clone(),
                    None => Box::pin(self.refresh_token())
                        .await
                        .context("failed to refresh token")?,
                };
                request = request.header("authorization", format!("token {token}"));
            }
        }
        Ok(request)
    }
}

struct GitHubInner {
    http: Client,
    org: String,
    client_id: String,
    private_key: PathBuf,
    token: Mutex<Option<GitHubToken>>,
}

#[derive(Clone)]
struct GitHubToken {
    token: String,
    issued_at: Instant,
}

pub struct CreatedRunner {
    pub id: u64,
    pub jitconfig: String,
}

#[derive(serde::Deserialize)]
pub struct RunnerInfo {
    pub busy: bool,
    pub status: String,
}

enum AuthMode<'a> {
    Jwt(&'a str),
    Token,
}

trait ResponseExt: Sized {
    async fn handle_github_error(self) -> Result<Self, Error>;
}

impl ResponseExt for Response {
    async fn handle_github_error(self) -> Result<Self, Error> {
        #[derive(serde::Deserialize)]
        struct ErrorResponse {
            message: String,
        }

        let status = self.status();
        if status.is_client_error() || status.is_server_error() {
            let url = self.url().to_string();
            let body = self.text().await?;
            let message = if let Ok(parsed) = serde_json::from_str::<ErrorResponse>(&body) {
                parsed.message
            } else {
                body
            };

            bail!("request failed with status {status}: {message} (url: {url})")
        } else {
            Ok(self)
        }
    }
}
